addConstraint, removeConstraint

harshithkashyap opened the pull request:
### Pull Request check-list
- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

### Description of change

Added `addConstraint`, `showConstraint`, `removeConstraint` methods. Closes #5212 

### Supported constraints

- UNIQUE
- CHECK
- DEFAULT (MSSQL only)
- PRIMARY KEY
- FOREIGN KEY

harshithkashyap commented:
@mickhansen @janmeier @sushantdhiman @felixfbecker Request review.
I've skipped tests for mysql & sqlite.
- Sqlite does not support add/remove constraints. 
- MySQL does not support `DROP CONSTRAINT` but it is rather implemented as `DROP PRIMARY`, - `DROP FOREIGN KEY FK_NAME` for primary and foreign keys.
- `DROP INDEX` should be used for `UNIQUE` constraint.
- `CHECK` constraints are silently ignored by the db engines

sushantdhiman commented:
@harshithkashyap 

> MySQL does not support DROP CONSTRAINT but it is rather implemented as DROP PRIMARY, DROP FOREIGN KEY FK_NAME for primary and foreign keys.
DROP INDEX should be used for UNIQUE constraint(MySQL).
CHECK constraints are silently ignored by the db engines(MySQL)

ORM should be able to abstract these changes, It wont help much if these methods cant be used for 50% dialects

> Sqlite does not support add/remove constraints.

You can create new table with constraints and move data to it

harshithkashyap commented:
@sushantdhiman I've added a `removeConstraint` method in mysql queryinterface and handle drop constraint based on the case.
SQLite stores the create table query in `sqlite_master` table. Since the specified constraints are not stored in any table/view, i've dropped down parsing and modifying sql queries.

harshithkashyap commented:
Ping @janmeier @mickhansen

mickhansen commented on docs/docs/migrations.md:
```diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
+queryInterface.addConstraint('Posts', ['username'], {
+  type: 'FOREIGN KEY',
+  references: { //Required field
+    model: 'target_table_name',
```
table

harshithkashyap replied:
@mickhansen I used the same naming conventions from [migration docs](http://docs.sequelizejs.com/en/v3/docs/migrations/). It has model/key for the references object.

sushantdhiman replied:
You can pass model here @harshithkashyap right, it will get table name from that ?

harshithkashyap replied:
I wasn't aware that we could pass model here. I'll modify the code to include the same.
Self Note: Add tests passing model

harshithkashyap replied:
@sushantdhiman Since constraints are executed only via migrations, i'll keep it to tableName only for now. Even `createTable` method expects a string and not a [model](https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/query-generator.js#L444-L467)

sushantdhiman replied:
@harshithkashyap Please rename it to `table` as @mickhansen suggested, that more suited if we dont support passing `model`

mickhansen commented on docs/docs/migrations.md:
```diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
+  type: 'FOREIGN KEY',
+  references: { //Required field
+    model: 'target_table_name',
+    key: 'target_column_name',
```
column? or field

mickhansen commented on test/unit/sql/remove-constraint.test.js:
```diff
@@ -0,0 +1,20 @@
+const sql = current.dialect.QueryGenerator;
+const dialect = Support.getTestDialect();
+
+if (dialect === 'mssql' || dialect === 'postgres') {
```
Can't remove constraints in mysql?
Please use dialect.features rather dialect naming checks

harshithkashyap replied:
No, mysql doesn't have `removeConstraint` in queryGenerator. We have to override `removeConstraint` in mysql queryInterface and it ultimately calls `QueryGenerator.removeIndexQuery` or `QueryGenerator.dropForeignKeyQuery` which already have tests covered

mickhansen commented on test/unit/sql/add-constraint.test.js:
```diff
@@ -0,0 +1,179 @@
+const expect = require('chai').expect;
+const dialect = Support.getTestDialect();
+
+if (dialect !== 'sqlite') {
```
features check, no constraints in sqlite?

harshithkashyap replied:
sqlite implementation of `addContraint` is a bit different. It has its own queryInterface method and queryGenerator functions. So, i've skipped them. But i'll go ahead and add unit tests for sqlite queryGenerator functions.

felixfbecker commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -560,6 +560,91 @@ const QueryGenerator = {
+      }
+    });
+
+    if(options.where) {
```
space before parens (saw this in multiple places)

harshithkashyap replied:
Thanks for catching. I'll fix all occurences and also add `keyword-spacing` to eslintrc

felixfbecker commented on lib/query-interface.js:
```diff
@@ -462,6 +462,51 @@ class QueryInterface {
+    options = Utils.cloneDeep(options);
+    options.fields = attributes;
+    if (this.sequelize.dialect.name === 'sqlite') {
+      return SQLiteQueryInterface.addConstraint.call(this, tableName, options, rawTablename);
```
I don't get this. The dialect-specific query interfaces inherit from the base one. So the user would get a SQLiteQueryInterface if they use SQLite. That may call `super`, in which case you get an infinite recursion here?
The children should call the parent, not the other way around

harshithkashyap replied:
@felixfbecker Only QueryGenerators are inherited. [`removeColumn`](https://github.com/sequelize/sequelize/blob/master/lib/query-interface.js#L333-L348) also uses the same pattern.

felixfbecker replied:
Okay

felixfbecker commented on lib/dialects/sqlite/query.js:
```diff
@@ -295,6 +302,50 @@ class Query extends AbstractQuery {
+      if (constraintSql.indexOf('REFERENCES') !== -1) {
+        //Parse out the constraint condition form sql string
+        update_action = constraintSql.match(/ON UPDATE (CASCADE|SET NULL|RESTRICT|NO ACTION|SET DEFAULT){1}/);
+        delete_action = constraintSql.match(/ON DELETE (CASCADE|SET NULL|RESTRICT|NO ACTION|SET DEFAULT){1}/);
```
why snake_case?

harshithkashyap replied:
Postgres & mssql returns it in snake case. Used the same here to maintain consistency

felixfbecker replied:
@janmeier @mickhansen wdyt?

harshithkashyap replied:
Apologies. Looks like MSSQL is the culprit here. I'm using MSSQL stored procedure [sp_helpconstraint](https://msdn.microsoft.com/en-us/library/ms176038.aspx) and hence can't format the results. I used the same snake_case for other dialects aswell.
Since default constraint in case of mssql are not part of ISO standard, i can't query on `INFORMATION_SCHEMA.TABLE_CONSTRAINTS` as we query in case of mysql and postgresql.
We could convert the snake_case to camelCase in mssql query.js and use sql aliases in case of mysql postgres?

sushantdhiman replied:
Postgres and MSSQL return format doesn't control variable name we use, does it ?

Just rename `update_action` => `updateAction` , `delete_action` => `deleteAction`

harshithkashyap replied:
Postgres is not a problem. Only MSSQL is. I'll convert snake_case keys to camelCase only in case of MSSQL.

harshithkashyap commented:
@sushantdhiman Can you take a look at this please?

sushantdhiman commented on docs/docs/migrations.md:
````diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
 queryInterface.removeIndex('Person', ['firstname', 'lastname'])
 ```
 
+### addConstraint(tableName, indexNameOrAttributes, options)
````
Why this is named as `indexNameOrAttributes`, I think it should be just `attributes`. Please explain your rationale behind this naming @harshithkashyap , I may be missing something

harshithkashyap replied:
Sorry. It should just be `attributes`. The signature in the code is correct. But i missed it here. Will change it

sushantdhiman commented on docs/docs/migrations.md:
```diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
+    roles: ['user', 'admin', 'moderator', 'guest']
+  }
+});
+//Constraint name is automatically constructed.
```
Remove this comment line

sushantdhiman commented on docs/docs/migrations.md:
```diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
+//UNIQUE
+queryInterface.addConstraint('Users', ['email'], {
+  type: 'unique',
+  name: 'custom_unique_constraint_name' //If not specifed, sequelize automatically creates a named constraint based on constraint type, table & column names
```
Remove the comment

sushantdhiman commented on docs/docs/migrations.md:
````diff
@@ -306,6 +306,74 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
 ```
 
+### addConstraint(tableName, indexNameOrAttributes, options)
+This method adds a new constraint of the specified type.
````
In the description explain meaning of arguments, and what common options we have. Here you can explain things like `options.name` and what happens if you dont pass them

sushantdhiman commented on lib/dialects/abstract/index.js:
```diff
@@ -42,7 +42,14 @@ AbstractDialect.prototype.supports = {
   constraints: {
-    restrict: true
+    restrict: true,
+    addConstraint: true,
```
why `addConstraint` and `dropConstraint` are part of this, is it for the fact some dialect doesn't support adding / dropping constraints

harshithkashyap replied:
Yes, Only MSSQL & postgres support complete `ADD CONSTRAINT` , `DROP CONSTRAINT` SQL keywords. MySQL supports `ADD CONSTRAINT` but there is no `DROP CONSTRAINT`. Sqlite doesn't support both of them. As @mickhansen suggested above, i used dialect.feature check instead of dialect.name checks to skip/allow tests

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+    if (options.where) {
+      options.where = this.whereItemsQuery(options.where);
+    }
+    const fieldsSqlQuotedString = fieldsSql.join(', ');
```
Empty line before this

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+    const fieldsSqlQuotedString = fieldsSql.join(', ');
+    const fieldsSqlString = fieldsSql.join('_');
+    let constraintSnippet, constraintName;
+    switch (options.type.toUpperCase()) {
```
Empty line before this

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+    switch (options.type.toUpperCase()) {
+      case 'UNIQUE':
+        constraintSnippet = `UNIQUE (${fieldsSqlQuotedString})`;
+        constraintName = `UK__${tableName}__${fieldsSqlString}`;
```
Use this format for naming `<table>_<column>_<constraint shorthand>`

sushantdhiman replied:
And no capital in shorthand

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        if (options.defaultValue === undefined) {
+          throw new Error('Default value must be specifed for DEFAULT CONSTRAINT');
+        }
+        if (this._dialect.name !== 'mssql') {
```
Empty line before this line

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        if (this._dialect.name !== 'mssql') {
+          throw new Error('Default constraints are supported only for MSSQL dialect.');
+        }
+        constraintSnippet = `DEFAULT (${this.escape(options.defaultValue)}) FOR ${fieldsSql[0]}`;
```
Empty line before this line

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        break;
+      case 'FOREIGN KEY':
+        const references = options.references;
+        if (!references || !references.model || !references.key) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        if (!references || !references.model || !references.key) {
+          throw new Error('references object with model and key must be specified');
+        }
+        const referencesSnippet = `${this.quoteTable(references.model)} (${this.quoteIdentifier(references.key)})`;
```
Empty line before this line

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        }
+        const referencesSnippet = `${this.quoteTable(references.model)} (${this.quoteIdentifier(references.key)})`;
+        constraintSnippet = `FOREIGN KEY (${fieldsSqlQuotedString}) REFERENCES ${referencesSnippet}`;
+        if (options.onUpdate) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        const referencesSnippet = `${this.quoteTable(references.model)} (${this.quoteIdentifier(references.key)})`;
+        constraintSnippet = `FOREIGN KEY (${fieldsSqlQuotedString}) REFERENCES ${referencesSnippet}`;
+        if (options.onUpdate) {
+          constraintSnippet+= ` ON UPDATE ${options.onUpdate.toUpperCase()}`;
```
`constraintSnippet +=` , space

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+          constraintSnippet+= ` ON UPDATE ${options.onUpdate.toUpperCase()}`;
+        }
+        if (options.onDelete) {
+          constraintSnippet+= ` ON DELETE ${options.onDelete.toUpperCase()}`;
```
`constraintSnippet +=` , space

sushantdhiman commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,91 @@ const QueryGenerator = {
+        if (options.onUpdate) {
+          constraintSnippet+= ` ON UPDATE ${options.onUpdate.toUpperCase()}`;
+        }
+        if (options.onDelete) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/mssql/query.js:
```diff
@@ -211,6 +213,11 @@ class Query extends AbstractQuery {
     });
   }
 
+  handleShowConstraintsQuery(data) {
```
No need for `handleShowConstraintsQuery` method, its just splicing data

harshithkashyap replied:
@sushantdhiman Keeping this as i've modified `handleShowConstraintsQuery` to include converting snake_case keys to camelCase.

sushantdhiman commented on lib/dialects/mssql/query.js:
```diff
@@ -256,6 +263,10 @@ class Query extends AbstractQuery {
+    match = err.message.match(/Could not drop constraint. See previous errors./);
+    if (match && match.length > 0) {
+      return new sequelizeErrors.UnknownConstraintError(match[1]);
+    }
```
Empty line after this line

sushantdhiman commented on lib/dialects/mssql/query.js:
```diff
@@ -256,6 +263,10 @@ class Query extends AbstractQuery {
     }
 
+    match = err.message.match(/Could not drop constraint. See previous errors./);
+    if (match && match.length > 0) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,15 @@ const QueryGenerator = {
   },
 
+  showConstraintsQuery(tableName, constraintName) {
+    let sql =  `SELECT CONSTRAINT_CATALOG AS constraint_catalog, CONSTRAINT_NAME AS constraint_name, CONSTRAINT_SCHEMA AS constraint_schema, CONSTRAINT_TYPE AS constraint_type, TABLE_NAME AS table_name, TABLE_SCHEMA AS table_schema from INFORMATION_SCHEMA.TABLE_CONSTRAINTS WHERE table_name='${tableName}'`;
```
Beautify this SQL statement, don't use one liner :cat2: , use template string

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,15 @@ const QueryGenerator = {
 
+  showConstraintsQuery(tableName, constraintName) {
+    let sql =  `SELECT CONSTRAINT_CATALOG AS constraint_catalog, CONSTRAINT_NAME AS constraint_name, CONSTRAINT_SCHEMA AS constraint_schema, CONSTRAINT_TYPE AS constraint_type, TABLE_NAME AS table_name, TABLE_SCHEMA AS table_schema from INFORMATION_SCHEMA.TABLE_CONSTRAINTS WHERE table_name='${tableName}'`;
+    if (constraintName) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,15 @@ const QueryGenerator = {
+  showConstraintsQuery(tableName, constraintName) {
+    let sql =  `SELECT CONSTRAINT_CATALOG AS constraint_catalog, CONSTRAINT_NAME AS constraint_name, CONSTRAINT_SCHEMA AS constraint_schema, CONSTRAINT_TYPE AS constraint_type, TABLE_NAME AS table_name, TABLE_SCHEMA AS table_schema from INFORMATION_SCHEMA.TABLE_CONSTRAINTS WHERE table_name='${tableName}'`;
+    if (constraintName) {
+      sql+= ` AND constraint_name = '${constraintName}'`;
```
space `sql +=`

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,15 @@ const QueryGenerator = {
+    if (constraintName) {
+      sql+= ` AND constraint_name = '${constraintName}'`;
+    }
+    sql+= ';';
```
Empty line before this line, space `sql +=`

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,15 @@ const QueryGenerator = {
+      sql+= ` AND constraint_name = '${constraintName}'`;
+    }
+    sql+= ';';
+    return sql;
```
Empty line before this line

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
 
+  getConstraintSnippet(tableName, options) {
+    let constraintSnippet, constraintName;
+    const fieldsSql = options.fields.map(field => {
```
Empty line before this line

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+        return result;
+      }
+    });
+    const fieldsSqlQuotedString = fieldsSql.join(', ');
```
Empty line before this line

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+    });
+    const fieldsSqlQuotedString = fieldsSql.join(', ');
+    const fieldsSqlString = fieldsSql.join('_');
+    switch (options.type.toUpperCase()) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+    const fieldsSqlString = fieldsSql.join('_');
+    switch (options.type.toUpperCase()) {
+      case 'UNIQUE':
+        constraintName = this.quoteIdentifier(options.name || `UK__${tableName}__${fieldsSqlString}`);      
```
Follow same constraint naming as suggested in `abstract/query_generator`

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+    switch (options.type.toUpperCase()) {
+      case 'UNIQUE':
+        constraintName = this.quoteIdentifier(options.name || `UK__${tableName}__${fieldsSqlString}`);      
+        constraintSnippet = `CONSTRAINT ${constraintName} UNIQUE (${fieldsSqlQuotedString})`;
```
This code section is almost duplicate of `abstract/query-gen` `addConstraint` method, can we use that method / minimize code duplication ?

harshithkashyap replied:
Yes, We can do that. I'll create a method to generate snippets which would be called from respective dialects

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+
+  showConstraintsQuery(tableName, constraintName) {
+    let sql =  `SELECT sql FROM sqlite_master WHERE tbl_name='${tableName}'`;
+    if (constraintName) {
```
Empty line before this line

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,74 @@ const QueryGenerator = {
+    if (constraintName) {
+      sql+= ` AND sql LIKE '%${constraintName}%'`;
+    }
+    sql+= ';';
```
Empty line before this line, space `sql += `

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -268,6 +340,29 @@ const QueryGenerator = {
       + `DROP TABLE ${quotedBackupTableName};`;
   },
 
+  alterConstraintQuery(tableName, attributes, createTableSql) {
```
This is a private method, right. I don think user will ever call this as it require `createTableSql`. It will be good idea to prefix such private methods with `_`

sushantdhiman commented on lib/query-interface.js:
```diff
@@ -462,6 +462,51 @@ class QueryInterface {
+      options = attributes;
+      attributes = options.fields;
+    }
+    if (!options.type) {
```
Empty line before this line

sushantdhiman commented on lib/query-interface.js:
```diff
@@ -462,6 +462,51 @@ class QueryInterface {
+    if (!options.type) {
+      throw new Error('Constraint type must be specified through options.type');
+    }
+    // testhint argsConform.end
```
whats this comment ? remove this comment

sushantdhiman commented on lib/query-interface.js:
```diff
@@ -462,6 +462,51 @@ class QueryInterface {
+
+    options = Utils.cloneDeep(options);
+    options.fields = attributes;
+    if (this.sequelize.dialect.name === 'sqlite') {
```
Empty line before this line

sushantdhiman commented:
@harshithkashyap what happens for SQLite case, does it support `addConstraint / removeConstraint`. How we are handling that ?

harshithkashyap commented:
@sushantdhiman No, SQLite doesn't support `addConstraint/removeConstraint`.
Though SQLite allows us to name constraints as described [here](https://www.sqlite.org/lang_createtable.html), SQLite does not store named constraints reference anywhere. We cannot removeConstraints by name.
The create table sql is stored in `sqlite_master` table. I'm querying on `sqlite_master` table, parsing the sql and adding constraint snippet at the end of create table query and recreating the table. In case of `removeConstraint`, we are finding the constraint snippet from the create table sql, remove it and recreate the table. 

There are other open issues as well regarding sqlite alter table https://github.com/sequelize/sequelize/issues/7078 and i could not find any other way to go about it.

harshithkashyap commented:
Ping @sushantdhiman @janmeier @felixfbecker

sushantdhiman commented on lib/dialects/mysql/query-generator.js:
```diff
@@ -193,6 +193,27 @@ const QueryGenerator = {
+
+    sql += ';';
+
+    return sql;
```
``return sql + ';';``

sushantdhiman commented on lib/dialects/sqlite/query-generator.js:
```diff
@@ -223,6 +223,18 @@ const QueryGenerator = {
+
+    sql += ';';
+    
+    return sql;
```
```return sql + ';';```

sushantdhiman commented on lib/dialects/sqlite/query-interface.js:
```diff
@@ -106,3 +107,63 @@ function renameColumn(tableName, attrNameBefore, attrNameAfter, options) {
+
+        createTableSql = createTableSql.replace(constraintSnippet, '');      
+        createTableSql += ';';
+        return this.describeTable(tableName, options);
```
New line above `return`, Please fix this for other sections that you have changed as well

felixfbecker commented on lib/dialects/abstract/query-generator.js:
```diff
@@ -561,6 +561,97 @@ const QueryGenerator = {
+    options = options || {};
+    const constraintSnippet = this.getConstraintSnippet(tableName, options);
+  
+    if (_.isString(tableName)) {
```
`typeof tableName === 'string'` is sufficient

felixfbecker commented on lib/dialects/mssql/query.js:
```diff
@@ -211,6 +213,20 @@ class Query extends AbstractQuery {
+  handleShowConstraintsQuery(data) {
+    data.splice(0, 1);
+    //Convert snake_case keys to camelCase as its generated by stored procedure
+    data = data.map(constraint => {
```
there is no point in using `map` here because you are mutating the argument

felixfbecker commented on lib/query-interface.js:
```diff
@@ -462,6 +462,53 @@ class QueryInterface {
+
+  showConstraint(tableName, options) {
+    const sql = this.QueryGenerator.showConstraintsQuery(tableName, options);
+    return this.sequelize.query(sql, _.assign({}, options, { type: QueryTypes.SHOWCONSTRAINTS }));
```
`Object.assign()`

harshithkashyap commented:
@felixfbecker @sushantdhiman I've pushed the suggested changes.

felixfbecker commented on lib/dialects/mssql/query.js:
```diff
@@ -211,6 +213,20 @@ class Query extends AbstractQuery {
+      for (const key in constraint) {
+        const camelKey = _.camelCase(key);
+        constraint[camelKey] = constraint[key];
+        delete constraint[key];
```
this will put the object in dictionary mode (disable V8 hidden classes). It is always faster to create a new object or set to `undefined`

harshithkashyap replied:
@felixfbecker  Setting it to undefined keeps both snake & camel cased keys to exist, though one of them would be undefined. I think its better to create a new object. Is this alright?
```javascript
handleShowConstraintsQuery(data) {
    data.splice(0, 1);
    //Convert snake_case keys to camelCase as its generated by stored procedure
    const constraints = data.map(result => {
      const constraint = {};
      for (const key in result) {
        const camelKey = _.camelCase(key);
        constraint[camelKey] = result[key];
      }
      return constraint;
    });

    return constraints;
  }
```

felixfbecker replied:
I would then make this function not mutate the argument:
```ts
handleShowConstraintsQuery(data) {
    // Convert snake_case keys to camelCase as its generated by stored procedure
    return data.slice(1).map(result => {
      const constraint = {};
      for (const key in result) {
        constraint[_.camelCase(key)] = result[key];
      }
      return constraint;
    });
  }
```

felixfbecker commented on lib/dialects/mysql/query-interface.js:
```diff
@@ -45,4 +46,27 @@ function removeColumn(tableName, columnName, options) {
+function removeConstraint(tableName, constraintName, options) {
+  /* jshint validthis:true */
+  const sql = this.QueryGenerator.showConstraintsQuery(tableName, constraintName);
+  return this.sequelize.query(sql, _.assign({}, options, { type: this.sequelize.QueryTypes.SHOWCONSTRAINTS }))
```
`Object.assign()`

sushantdhiman commented:
@harshithkashyap Please rebase this branch and take care of my last comment (Using `table` instead of `model`) :) We should merge it, its a good feature for CLI as well

harshithkashyap commented on docs/docs/migrations.md:
```diff
@@ -306,6 +306,84 @@ queryInterface.removeIndex('Person', 'SuperDuperIndex')
+  references: { //Required field
+    model: 'target_table_name',
+    key: 'target_column_name'
+  },
```
@sushantdhiman Since we are changing `model` to `table`, would it better to rename `key` to `column` as well?

sushantdhiman replied:
Probably `field` will make sense, as in our code base `column` is usually represented as `field`

sushantdhiman commented:
@harshithkashyap Great work ❤️

yuanfeiz commented:
Which version will include this patch?

felixfbecker commented:
@yuanfeiz the next ;)

yuanfeiz commented:
Cool! Timely deliver!
